File: Intermediate Challenges/Challenge #0050 [Intermediate]/solutions/solution.c
#include <stdio.h>
#include <stdlib.h>

#include <getopt.h>
#include <sys/types.h>
#include <dirent.h>
#include <unistd.h>

static char *invoc_name = NULL;
static char *version_string = "1.0";

static struct option longopts[] = {
    { "path", required_argument, NULL, 'p' },
    { "depth", required_argument, NULL, 'd' },
    { "help", no_argument, NULL, 'h' },
    { "version", no_argument, NULL, 'v' },
    { NULL, 0, NULL, 0 },
};


static void usage(void);
static void version(void);

static void print_tree(char *directory, int depth, char *preface);

int main(int argc, char **argv){
    int c;
    int depth = -1;
    char *base_dir = NULL;
    invoc_name = argv[0];
    while ( ( c = getopt_long(argc, argv, "p:d:hv", longopts, NULL) ) != -1 ) {
        switch (c) {
            case 'p':
                base_dir = optarg;
                break;
            case 'd':
                c = sscanf(optarg, "%d", &depth);
                if ( c != 1 ) {
                    fprintf(stderr, "Invalid depth: %s\n", optarg);
                    fprintf(stderr, "Try %s --help for more information.\n", invoc_name);
                    exit(EXIT_FAILURE);
                }
                break;
            case 'h':
                usage();
                exit(EXIT_SUCCESS);
            case 'v':
                version();
                exit(EXIT_SUCCESS);
            case '?':
                fprintf(stderr, "Try %s --help for more information.\n", invoc_name);
                exit(EXIT_FAILURE);
            default:
                break;
        }
    }

    /* default to current directory */
    if ( base_dir == NULL ) {
        base_dir = malloc(sizeof(char) * 2);
        base_dir[0] = '.';
        base_dir[1] = '\0';
    }

    //print_tree(base_dir, depth, NULL);

    return 0;
}

static void print_tree(char *directory, int depth, char *preface){
    DIR *curr_dir;
    if ( fchdir(fd) ) {
        perror("Cannot use directory");
        exit(EXIT_FAILURE);
    }

    curr_dir = fopendir(fd);
#if 0
    DIR *curr_dir = NULL;
    chdir(directory);
    /* check whether successful */
    curr_dir = opendir(".");
    if (root_dir == NULL) {
        perror("Cannot open directory");
        return;
    }

    if ( preface == NULL )
        printf("  ");

    printf("%s", directory);
    /* if there are directories too */
    if ( depth != 0 )
        printf("%s  +", directory);

    closedir(curr_dir);
#endif

    closedir(curr_dir);

    chdir("..");
}

static void usage(void){
    printf( "Usage: %s [LONG-OPTION]...", invoc_name);
    printf( "   or  %s [SHORT-OPTION]...", invoc_name);
    printf( "Prints out a tree for a given directory or, if none given, the current directory.");
    printf( "");
    printf( "   -p, --path      create tree of this directory");
    printf( "   -d, --depth     show tree to a depth of d");
    printf( "   -h, --help      display this help and exit");
    printf( "   -v, --version   display version information and exit");
}

static void version(void){
    printf("Directory Tree v%s\n", version_string);
}

File: Intermediate Challenges/Challenge #0050 [Intermediate]/solutions/solution.pl
#!/usr/bin/perl --

use v5.18;
use warnings;
use utf8;
use Getopt::Long;

binmode STDOUT, ":utf8";

my $version_string = "1.0";
my $depth = -1;
my $base_dir = '.';

GetOptions (
             "depth"   => \$depth,
             "path"    => \$base_dir,
             "help"    => \&usage,
             "version" => \&version,
           )
           or die "Try $0 --help for more information";

print_directory_tree($base_dir, $depth, "");

sub print_directory_tree{
    my ($directory, $depth, $preface) = @_;
    chdir $directory or warn "Here";
    opendir my $dh, "." or die "Cannot open $directory: $!";
    my @dirs = sort grep { -d and not /^\./ } readdir($dh);
    closedir $dh;

    print "  " if $preface eq "";
    say "$directory";
    say "$preface  +" if @dirs and $depth != 0;

    # For each directory
    for ( my $i = 0; $i < @dirs; $i++ ) {
        my $dir = $dirs[$i];
        if ( $depth != 0 ) { # if depth < 0 then as far as can go.
            say "$preface  |";
            print "$preface  +--";
            # if not last sub-directory then should extend preface
            # so that sub-subs are further across
            print_directory_tree($dir, $depth - 1, ($i < $#dirs) ? "$preface  |" : "$preface   ")
        }
    }
    chdir "..";
}

sub usage{
    say "Usage: $0 [LONG-OPTION]...";
    say "   or  $0 [SHORT-OPTION]...";
    say "Displays a tree for each DIRECTORY";
    say "";
    say "   -p, --path      create tree of this directory";
    say "   -d, --depth     show tree to a depth of d";
    say "   -h, --help      display this help and exit";
    say "   -v, --version   display version information and exit";
    exit;
}

sub version{
    say "Tree v$version_string";
    exit;
}
